File: haskell/ch1/Ex117.hs
-- :cd C:\Users\Yohan\IdeaProjects\sicp\haskell\ch1
-- :load ExXXX
-- Run with func_name arg1...

module Ex117 where

mult_invariant _ 0 = 0
mult_invariant a b | even b = 2 * (mult_invariant a (div b 2))
mult_invariant a b = a + mult_invariant a (b - 1)

times a b = mult_invariant a b

File: javascript/ch1/Ex117.js
// Load with: var ExXXX = require('./ExXXX')
// Run with: ExXXX.xxx()

even = function (x) {
    return x % 2 == 0;
};

mult_invariant = function (a, b) {
    if (b == 0) return 0;
    if (even(b)) return 2 * mult_invariant(a, b / 2);
    return a + mult_invariant(a, b - 1);
};

times = function (a, b) {
    return mult_invariant(a, b);
};


exports.times = times;